doc/threadsafe_function_ex.md

```diff
@@ -0,0 +1,317 @@
+```
+
+Returns one of:
+- `napi_ok`: The thread-safe function has been successfully released.
+- `napi_invalid_arg`: The thread-safe function's thread-count is zero.
+- `napi_generic_failure`: A generic error occurred when attemping to release the
+  thread-safe function.
+
+### Abort
+
+"Abort" the thread-safe function. This will cause all subsequent APIs associated
+with the thread-safe function except `Release()` to return `napi_closing` even
+before its reference count reaches zero. In particular, `BlockingCall` and
+`NonBlockingCall()` will return `napi_closing`, thus informing the threads that
+it is no longer possible to make asynchronous calls to the thread-safe function.
+This can be used as a criterion for terminating the thread. Upon receiving a
+return value of `napi_closing` from a thread-safe function call a thread must
+make no further use of the thread-safe function because it is no longer
+guaranteed to be allocated.
+
+```cpp
+napi_status Napi::ThreadSafeFunctionEx<ContextType, DataType, Callback>::Abort()
+```
+
+Returns one of:
+- `napi_ok`: The thread-safe function has been successfully aborted.
+- `napi_invalid_arg`: The thread-safe function's thread-count is zero.
+- `napi_generic_failure`: A generic error occurred when attemping to abort the
+  thread-safe function.
+
+### BlockingCall / NonBlockingCall
+
+Calls the Javascript function in either a blocking or non-blocking fashion.
+- `BlockingCall()`: the API blocks until space becomes available in the queue.
+  Will never block if the thread-safe function was created with a maximum queue
+  size of `0`.
+- `NonBlockingCall()`: will return `napi_queue_full` if the queue was full,
+  preventing data from being successfully added to the queue.
+
+```cpp
+napi_status Napi::ThreadSafeFunctionEx<ContextType, DataType, Callback>::BlockingCall(DataType* data = nullptr) const
+
+napi_status Napi::ThreadSafeFunctionEx<ContextType, DataType, Callback>::NonBlockingCall(DataType* data = nullptr) const
+```
+
+- `[optional] data`: Data to pass to the callback which was passed to
+  `ThreadSafeFunctionEx::New()`.
+
+Returns one of:
+- `napi_ok`: The call was successfully added to the queue.
+- `napi_queue_full`: The queue was full when trying to call in a non-blocking
+  method.
+- `napi_closing`: The thread-safe function is aborted and cannot accept more
+  calls.
+- `napi_invalid_arg`: The thread-safe function is closed.
+- `napi_generic_failure`: A generic error occurred when attemping to add to the
+  queue.
+
+
+## Example
+
+```cpp
+#include <chrono>
+#include <thread>
+#include <napi.h>
+
+using namespace Napi;
+
+using Context = Reference<Value>;
+using DataType = int;
+void CallJs( Napi::Env env, Function callback, Context* context, DataType* data );
+using TSFN = ThreadSafeFunctionEx<Context, DataType, CallJs>;
+using FinalizerDataType = void;
+
+std::thread nativeThread;
+TSFN tsfn;
+
+Value Start( const CallbackInfo& info )
```

Comment:
Our coding style has no spaces after brackets and the opening brace is one space away from the last character on the line, not on the next line.

Reply:
Do we have some sort of clang-format file for formatting...?